File: internal/codegen/templates/runtime.py
_T = TypeVar("_T", bound=Message)

@dataclass(frozen=True)
class Evaluation(Generic[_T]):
    value: _T
    # Original SDK response: answers, probabilities, confidence, model and usage.
    response: Any


def _questions(rules: list[dict[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for q in rules:
        if q["type"] == "choice":
            result[q["name"]] = Choice(instructions=q["instructions"], criteria=q["choices"])
        elif q["type"] == "score":
            result[q["name"]] = Score(instructions=q["instructions"], criteria=[l["description"] for l in q["levels"]])
        else:
            result[q["name"]] = Noul(instructions=q["instructions"])
    return result


def _get(obj: Any, key: str, default: Any = None) -> Any:
    return obj.get(key, default) if isinstance(obj, dict) else getattr(obj, key, default)


def _map_response(response: Any, rules: list[dict[str, Any]]) -> dict[str, Any]:
    values: dict[str, Any] = {}
    sentinel = object()
    answers = _get(response, "answers", sentinel)
    canonical = answers is not sentinel
    for q in rules:
        group = answers if canonical else _get(response, {"choice":"choices", "noul":"nouls", "score":"scores"}[q["type"]])
        if not isinstance(group, dict) or q["name"] not in group:
            raise ValueError(f"Missing answer for {q['name']}")
        item = group[q["name"]]
        kind = _get(item, "type", sentinel)
        if (canonical or kind is not sentinel) and kind != q["type"]:
            raise ValueError(f"{q['name']}: expected answer type {q['type']}")
        if q["type"] == "choice":
            label = _get(item, "choice")
            if not isinstance(label, str) or label not in q["choices"]:
                raise ValueError(f"{q['name']}: invalid choice")
            if q.get("oneof"):
                values[label] = base64.b64encode(label.encode("ascii")).decode("ascii") if q["oneof"][label] == "bytes" else label
            else:
                values[q["field"]] = label
        elif q["type"] == "noul":
            value = _get(item, "noul", sentinel)
            if value is sentinel:
                value = _get(item, "result")
            if isinstance(value, bool):
                values[q["field"]] = value
            elif isinstance(value, (int, float)) and math.isfinite(value) and 0 <= value <= 1:
                values[q["field"]] = value >= q["threshold"]
            else:
                raise ValueError(f"{q['name']}: invalid noul")
        else:
            pos = _get(item, "score")
            levels = q["levels"]
            if isinstance(pos, bool) or not isinstance(pos, (int, float)) or not math.isfinite(pos) or not 0 <= pos <= len(levels)-1:
                raise ValueError(f"{q['name']}: invalid score")
            i = int(pos)
            value = float(levels[i]["value"])
            if i + 1 < len(levels):
                value = (1 - (pos - i)) * value + (pos - i) * levels[i + 1]["value"]
            if not math.isfinite(value):
                raise ValueError(f"{q['name']}: non-finite domain value")
            if q["kind"].startswith(("int", "uint")):
                whole = math.floor(abs(value))
                value = math.copysign(whole + (1 if abs(value) - whole >= 0.5 else 0), value)
            values[q["field"]] = str(int(value)) if q["kind"] in ("int64", "uint64") else value
    return values

File: internal/codegen/templates/runtime.ts
interface Rule {
  name: string;
  type: "choice" | "noul" | "score";
  instructions: string;
  field: string;
  kind: string;
  choices?: Record<string, string>;
  levels?: { value: number; description: string }[];
  threshold: number;
  oneof?: Record<string, string>;
}

export interface Evaluation<T> {
  value: T;
  /** Complete SDK response, including answers, probabilities, confidence, model and usage. */
  response: Record<string, unknown>;
}

function object(value: unknown, label: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label}: expected an object`);
  }
  return value as Record<string, unknown>;
}

function scoreQuestion(q: Rule) {
  const descriptions = (q.levels ?? []).map(l => l.description);
  if (descriptions.length < 2) throw new Error(`${q.name}: at least two score levels are required`);
  return score(q.instructions, [descriptions[0], descriptions[1], ...descriptions.slice(2)]);
}

function questions(rules: readonly Rule[]) {
  return Object.fromEntries(rules.map(q => [q.name,
    q.type === "choice" ? choice(q.instructions, q.choices ?? {}) :
    q.type === "score" ? scoreQuestion(q) :
    noul(q.instructions)
  ]));
}

function mapResponse(response: Record<string, unknown>, rules: readonly Rule[]): JsonObject {
  const values: JsonObject = Object.create(null) as JsonObject;
  const canonical = Object.hasOwn(response, "answers");
  for (const q of rules) {
    const group = object(response[canonical ? "answers" : {choice:"choices", noul:"nouls", score:"scores"}[q.type]], q.name);
    if (!Object.hasOwn(group, q.name)) throw new Error(`Missing answer for ${q.name}`);
    const item = object(group[q.name], q.name);
    if ((canonical || Object.hasOwn(item, "type")) && item.type !== q.type) {
      throw new Error(`${q.name}: expected answer type ${q.type}`);
    }
    if (q.type === "choice") {
      const label = item.choice;
      if (typeof label !== "string" || !Object.hasOwn(q.choices ?? {}, label)) {
        throw new Error(`${q.name}: invalid choice`);
      }
      if (q.oneof) {
        if (q.oneof[label] === "bytes") {
          // Protobuf field identifiers are ASCII, so the selected label needs no Unicode conversion.
          values[label] = btoa(label);
        } else { values[label] = label; }
      } else { values[q.field] = label; }
    } else if (q.type === "noul") {
      const value = Object.hasOwn(item, "noul") ? item.noul : item.result;
      if (typeof value === "boolean") { values[q.field] = value; }
      else if (typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1) {
        values[q.field] = value >= q.threshold;
      } else { throw new Error(`${q.name}: invalid noul`); }
    } else {
      const pos = item.score;
      const levels = q.levels ?? [];
      if (typeof pos !== "number" || !Number.isFinite(pos) || levels.length < 2 || pos < 0 || pos > levels.length - 1) {
        throw new Error(`${q.name}: invalid score`);
      }
      const i = Math.floor(pos);
      let value = levels[i].value;
      if (i + 1 < levels.length) value = (1 - (pos - i)) * value + (pos - i) * levels[i + 1].value;
      if (q.kind.startsWith("int") || q.kind.startsWith("uint")) {
        const magnitude = Math.abs(value);
        const whole = Math.floor(magnitude);
        value = Math.sign(value) * (whole + (magnitude - whole >= 0.5 ? 1 : 0));
      }
      if (!Number.isFinite(value)) throw new Error(`${q.name}: non-finite domain value`);
      values[q.field] = q.kind === "int64" || q.kind === "uint64" ? value.toFixed(0) : q.kind === "float32" ? Math.fround(value) : value;
    }
  }
  return values;
}
